src/rapids_singlecell/preprocessing/_harmony/__init__.py

```diff
@@ -0,0 +1,446 @@
+
+    tol_clustering
+        Tolerance on justifying convergence of the clustering step over objective function values within each Harmony iteration.
+
+    ridge_lambda
+        Hyperparameter of ridge regression on the correction step.
+
+    sigma
+        Weight of the entropy term in objective function.
+
+    block_proportion
+        Proportion of block size in one update operation of clustering step.
+
+    theta
+        Weight of the diversity penalty term in objective function.
+
+    tau
+        Discounting factor on ``theta``. By default, there is no discounting.
+
+    correction_method
+        Choose which method for the correction step: ``original`` for original method, ``fast`` for improved method. By default, use improved method.
+
+    random_state
+        Random seed for reproducing results.
+
+    verbose
+        If ``True``, print verbose output.
+
+    Returns
+    -------
+    The integrated embedding by Harmony, of the same shape as the input embedding.
+
+    Examples
+    --------
+    >>> adata = anndata.read_h5ad("filename.h5ad")
+    >>> X_harmony = harmonize(adata.obsm['X_pca'], adata.obs, 'Channel')
+
+    >>> adata = anndata.read_h5ad("filename.h5ad")
+    >>> X_harmony = harmonize(adata.obsm['X_pca'], adata.obs, ['Channel', 'Lab'])
+    """
+
+    Z_norm = _normalize_cp(Z)
+    n_cells = Z.shape[0]
+
+    batch_codes = _get_batch_codes(batch_mat, batch_key)
+    n_batches = batch_codes.cat.categories.size
+    N_b = cp.array(batch_codes.value_counts(sort=False).values, dtype=Z.dtype)
+    Pr_b = (N_b.reshape(-1, 1) / len(batch_codes)).astype(Z.dtype)
+
+    Phi = _one_hot_tensor_cp(batch_codes).astype(Z.dtype)
+    if n_clusters is None:
+        n_clusters = int(min(100, n_cells / 30))
+    theta = (cp.ones(n_batches) * theta).astype(Z.dtype)
+
+    if tau > 0:
+        theta = theta * (1 - cp.exp(-N_b / (n_clusters * tau)) ** 2)
+
+    theta = theta.reshape(1, -1)
+    assert block_proportion > 0 and block_proportion <= 1
+    assert correction_method in {"fast", "original"}
+
+    cp.random.seed(random_state)
+
+    # Initialize centroids
+    R, E, O, objectives_harmony = _initialize_centroids(
+        Z_norm,
+        n_clusters,
+        sigma,
+        Pr_b,
+        Phi,
+        theta,
+        random_state,
+    )
+    if verbose:
+        print("Initialization is completed.")
+
+    for i in range(max_iter_harmony):
+        _clustering(
+            Z_norm,
+            Pr_b,
+            Phi,
+            R,
+            E,
+            O,
+            n_clusters,
+            theta,
+            tol_clustering,
+            objectives_harmony,
+            max_iter_clustering,
+            sigma,
+            block_proportion,
+        )
+
+        Z_hat = _correction(Z, R, Phi, O, ridge_lambda, correction_method)
+        Z_norm = _normalize_cp(Z_hat, p=2)
+        if verbose:
+            print(f"\tCompleted {i + 1} / {max_iter_harmony} iteration(s).")
```

Comment:
don’t you have some logging infra you should use instead?

I don’t want to see a single `print` statement in any library I use (if it has a CLI, that one may have print statements)